js_modules/dagit/packages/ui/src/components/TagSelector.tsx

```diff
@@ -0,0 +1,170 @@
+  selected: boolean;
+};
+type Props = {
+  placeholder?: React.ReactNode;
+  allTags: string[];
+  selectedTags: string[];
+  setSelectedTags: (tags: React.SetStateAction<string[]>) => void;
+  renderTag?: (tag: string, tagProps: TagProps) => React.ReactNode;
+  renderTagList?: (tags: React.ReactNode[]) => React.ReactNode;
+  renderDropdown?: (dropdown: React.ReactNode) => React.ReactNode;
+  renderDropdownItem?: (tag: string, dropdownItemProps: DropdownItemProps) => React.ReactNode;
+  dropdownStyles?: React.CSSProperties;
+};
+
+const defaultRenderTag = (tag: string, tagProps: TagProps) => {
+  return (
+    <Tag>
+      <Box flex={{direction: 'row', gap: 4, justifyContent: 'space-between', alignItems: 'center'}}>
+        <span>{tag}</span>
+        <Box style={{cursor: 'pointer'}} onClick={tagProps.close}>
+          <Icon name="close" />
+        </Box>
+      </Box>
+    </Tag>
+  );
+};
+
+const defaultRenderDropdownItem = (tag: string, dropdownItemProps: DropdownItemProps) => {
+  return (
+    <label>
+      <MenuItem
+        text={
+          <Box flex={{alignItems: 'center', gap: 8}}>
+            <Checkbox checked={dropdownItemProps.selected} onChange={dropdownItemProps.toggle} />
+            <span>{tag}</span>
+          </Box>
+        }
+      />
+    </label>
+  );
+};
+
+export const TagSelector = ({
+  allTags,
+  placeholder,
+  selectedTags,
+  setSelectedTags,
+  renderTag,
+  renderDropdownItem,
+  renderDropdown,
+  dropdownStyles,
+  renderTagList,
+}: Props) => {
+  const [isDropdownOpen, setIsDropdownOpen] = React.useState(false);
+  const dropdown = React.useMemo(() => {
+    const dropdownContent = (
+      <Box
+        style={{
+          maxHeight: '500px',
+          overflowY: 'auto',
+          ...dropdownStyles,
+        }}
+      >
+        {allTags.map((tag) => {
+          const selected = selectedTags.includes(tag);
+          const toggle = () => {
+            setSelectedTags(
+              selected ? selectedTags.filter((t) => t !== tag) : [...selectedTags, tag],
+            );
+          };
+          if (renderDropdownItem) {
+            return <div key={tag}>{renderDropdownItem(tag, {toggle, selected})}</div>;
+          }
+          return defaultRenderDropdownItem(tag, {toggle, selected});
+        })}
+      </Box>
+    );
+    if (renderDropdown) {
+      return renderDropdown(dropdownContent);
+    }
+    return <Menu>{dropdownContent}</Menu>;
+  }, [allTags, dropdownStyles, renderDropdown, renderDropdownItem, selectedTags, setSelectedTags]);
+
+  const dropdownContainer = React.useRef<HTMLDivElement>(null);
+
+  const tagsContent = React.useMemo(() => {
+    if (selectedTags.length === 0) {
+      return <Placeholder>{placeholder || 'Select tags'}</Placeholder>;
+    }
+    const tags = selectedTags.map((tag) =>
+      (renderTag || defaultRenderTag)(tag, {
+        close: (ev) => {
+          setSelectedTags((tags) => tags.filter((t) => t !== tag));
+          ev.stopPropagation();
```

Comment:
Why?

Reply:
because I have an `onClick` on the parent container to open the dropdown.